File: 백준/Gold/14267. 회사 문화 1/회사 문화 1.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
using ll = long long;
using pii = pair<int, int>;
using vi = vector<int>;
const int MAX = 100001;
int N, M;
vi tree[MAX];
int prs[MAX];
int dp[MAX];

void dfs(int now) {
	for (int nxt : tree[now]) {
		if (prs[now] == nxt) continue;
		dp[nxt] += dp[now];
		dfs(nxt);
	}
}

int main() {
    cin.tie(0)->sync_with_stdio(0);
	cin >> N >> M;

	for(int i=1; i<=N; ++i) {
		int num;
		cin >> num;
		prs[i] = num;
		if (num != -1) {
			tree[num].push_back(i);
			tree[i].push_back(num);
		}
	}
	for(int i=0; i<M; ++i) {
		int a, b;
		cin >> a >> b;
		dp[a] += b;		
	}
	dfs(1);
	for(int i=1; i<=N; ++i) cout << dp[i] << " ";
}

File: 백준/Gold/14267. 회사 문화 1/회사 문화 1.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**7)
MAX = 100001
prs = [0 for _ in range(MAX)]
dp = [0 for _ in range(MAX)]
tree = [[] for _ in range(MAX)]
arr = [0]

N, M = map(int, input().split())
arr += list(map(int, input().split()))

for i in range(1, N+1):
    if arr[i] != -1:
        tree[arr[i]].append(i)

for i in range(M):
    a, b = map(int, input().split())
    dp[a] += b

def dfs(now):
    for nxt in tree[now]:
        if prs[now] == nxt:
            continue
        dp[nxt] += dp[now]
        dfs(nxt)
    
dfs(1)
for i in range(1, N+1):
    print(dp[i], end=" ")